src/dicomweb_client/uri.py

```diff
@@ -0,0 +1,358 @@
+import urllib.parse as urlparse
+
+
+class URIType(enum.Enum):
+    """Type of DICOM resource the URI points to."""
+    SERVICE = 'service'
+    STUDY = 'study'
+    SERIES = 'series'
+    INSTANCE = 'instance'
+
+
+# For DICOM Standard spec validation of UID components in `URI`.
+_MAX_UID_LENGTH = 64
+_REGEX_UID = re.compile(r'[0-9]+([.][0-9]+)*')
+
+
+class URI:
+    """Class to represent a fully qualified HTTP[S] URI to a DICOMweb resource.
+
+    http://dicom.nema.org/dicom/2013/output/chtml/part18/sect_6.7.html
+
+    Given an HTTP[S] `base_url`, a valid DICOMweb-compatible URI would be:
+
+    - ``<base_url>`` (no DICOMWeb suffix)
+    - ``<base_url>/studies/<study_instance_uid>``
+    - ``<base_url>/studies/<study_instance_uid>/series/<series_instance_uid>``
+    - ``<base_url>/studies/<study_instance_uid>/series/<series_instance_uid>/instances/<sop_instance_uid>``
+    """
+
+    def __init__(self,
+                 base_url: str,
+                 study_instance_uid: Optional[str] = None,
+                 series_instance_uid: Optional[str] = None,
+                 sop_instance_uid: Optional[str] = None):
+        """Instantiates an object.
+
+        As per the DICOM Standard, the Study, Series, and Instance UIDs must be
+        a series of numeric components (``0``-``9``) separated by the period
+        ``.`` character, with a maximum length of 64 characters.
+
+        Parameters
+        ----------
+        base_url: str
+            DICOMweb service HTTP[S] URL. Trailing forward slashes are not
+            permitted.
+        study_instance_uid: str, optional
+            DICOM Study Instance UID.
+        series_instance_uid: str, optional
+            DICOM Series Instance UID.
+        sop_instance_uid: str, optional
+            DICOM SOP Instance UID.
+
+        Raises
+        ------
+        ValueError
+            In the following cases:
+
+            - `base_url` has a trailing slash.
+            - `base_url` does not use the HTTP[S] addressing scheme.
+            - `base_url` is incompatible with the DICOMweb standard.
+            - `series_instance_uid` is supplied without `study_instance_uid`.
+            - `sop_instance_uid` is supplied without `study_instance_uid` or
+              `series_instance_uid`.
+            - Any one of `study_instance_uid`, `series_instance_uid`, or
+              `sop_instance_uid` does not meet the DICOM Standard UID spec in
+              the docstring.
+        """
+        _validate_base_url(base_url)
+        _validate_uids(study_instance_uid, series_instance_uid,
+                       sop_instance_uid)
+        self._base_url = base_url
+        self._study_instance_uid = study_instance_uid
+        self._series_instance_uid = series_instance_uid
+        self._instance_uid = sop_instance_uid
+
+    def __str__(self) -> str:
+        """Returns the object as a DICOMweb URI string."""
+        parts = (('studies', self.study_instance_uid),
```

Comment:
For multiline statements (tuples, lists, strings, etc.), I generally prefer the following style

```python
parts = (
    ('studies', self.study_instance_uid),
    ('series', self.series_instance_uid),
    ('instances', self.sop_instance_uid),
)
```

with one item per line. This significantly improves readability in my opinion.

Reply:
Agreed, done.